driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -337,28 +337,43 @@ private PoolingOptions options() {
       }
     }
 
+    Connection leastBusy = null;
+
     if (connections[shardId].isEmpty()) {
       if (host.convictionPolicy.canReconnectNow()) {
         if (connectionsPerShard == 0) {
           maybeSpawnNewConnection(shardId);
+          return enqueue(timeout, unit, maxQueueSize, shardId);
         } else if (scheduledForCreation[shardId].compareAndSet(0, connectionsPerShard)) {
           for (int i = 0; i < connectionsPerShard; i++) {
             // We don't respect MAX_SIMULTANEOUS_CREATION here because it's  only to
             // protect against creating connection in excess of core too quickly
             manager.blockingExecutor().submit(new ConnectionTask(shardId));
           }
+          return enqueue(timeout, unit, maxQueueSize, shardId);
         }
-        return enqueue(timeout, unit, maxQueueSize, shardId);
       }
-    }
-
-    int minInFlight = Integer.MAX_VALUE;
-    Connection leastBusy = null;
-    for (Connection connection : connections[shardId]) {
-      int inFlight = connection.inFlight.get();
-      if (inFlight < minInFlight) {
-        minInFlight = inFlight;
-        leastBusy = connection;
+      // connections for this shard are still being initialized so pick connection for any shard
+      int minInFlight = Integer.MAX_VALUE;
+      for (int index = (shardId + 1) % connections.length;
+          index != shardId;
+          index = (index + 1) % connections.length) {
+        for (Connection connection : connections[index]) {
+          int inFlight = connection.inFlight.get();
+          if (inFlight < minInFlight) {
+            minInFlight = inFlight;
+            leastBusy = connection;
+          }
+        }
+      }
+    } else {
+      int minInFlight = Integer.MAX_VALUE;
+      for (Connection connection : connections[shardId]) {
+        int inFlight = connection.inFlight.get();
+        if (inFlight < minInFlight) {
+          minInFlight = inFlight;
+          leastBusy = connection;
+        }
```

Comment:
You could turn the above into a function and reuse it above, to please the `$DEITY` of clean code.